Fix Move regression

reaperrr opened the pull request:
If `progress == Distance`, we must not move again on the same tick, but still `return true` to avoid losing a tick in the case
when this is the last `Move` tick followed by a different activity (or a new queued `Move`, for example via waypoints).

Fixes #19347.

reaperrr committed:
Fix Move regression

If progress == Distance, we must not move again on the same tick,
but still 'return true' to avoid losing a tick in the case
when this is the last Move tick followed by a different activity
(or a new queued Move, for example via waypoints).

pchote commented:
This doesn't appear to be sufficient to solve the issue, or maybe i'm doing something silly.

Testcase:
* Apply rule changes from 8ef025a283c8f861f4c7a04623e07bd98a3ca9cb
* Line up 1tnk, 2tnk, 3tnk next to eachother facing a long strip of clear terrain
* Pause the game and use the `/terrainoverlay` to issue move orders in a straight line
* Unpause and watch the three tanks race

The Medium Tank surges ahead with noticable jerks showing that it is still double-ticking.

reaperrr commented:
Updated.

Fingers crossed, but I think I got it. Passes the above testcase now, at least.

Changed the approach to tracking the world tick we're on, which I believe makes for cleaner and easier-to-understand code, too.

pchote commented:
Good idea. I think we might now be able to remove the `firstTick` stuff and simplify the comment too?

reaperrr commented:
Updated, but put the changes in a separate commit for now.

pchote commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -442,20 +441,18 @@ public override bool Tick(Actor self)
 					mobile.Facing = ToFacing;
 
 					Queue(OnComplete(self, mobile, Move));
+					Move.lastMovePartCompletedTick = self.World.WorldTick;
```
Minor nit: It would be more robust to set this before calling `OnComplete`, as this makes sure that any future code changes that might read `Move.lastMovePartCompletedTick` from the `MovePart` ctor see the same value.

pchote commented:
LGTM otherwise and I didn't see any problems with a quick test ingame after making the above change. So :+1: after addressing that and squashing.

reaperrr commented:
Updated.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5c95f2e0b3235c1949f6e3fc62405ba5a9fc1f99...fb1536336f5684e15eccc961ecf367f0226f5950
